### PR TITLE
[BAQE-1259] Create dedicated LDAP instance for each Openshift project

### PR DESCRIPTION
**Pull request by Sgitario**
Verified using 7.7.0 latest image from 20200210 build date using templates and operators.

### CODE REVIEW

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    public String getLdapDockerImage() {
+        return getProperty(LDAP_DOCKER_IMAGE_PROPERTY);
+    }
+
+    public String getProperty(String key) {
+        return prop.getProperty(key);
+    }
+
+    public synchronized static final CloudProperties getInstance() {
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Can you please adjust order of keywords?
The codebase moreless follows:
https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.4.3
https://stackoverflow.com/questions/10299067/modifier-keyword-order-in-java

File: framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/deployment/LdapDeployment.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.cloud.api.deployment;
+
+/**
+ * Maven repository deployment representation in cloud.
```
Comment (sutaakar) [inline framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/deployment/LdapDeployment.java]:
```suggestion
 * LDAP deployment representation in cloud.
```

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        if (serviceName == null) {
+            serviceName = ServiceUtil.getLdapServiceName(getOpenShift());
+        }
+        return serviceName;
+    }
+
+    @Override
+    public String getHost() {
+        if (host == null) {
+            host = getLdapRoute(getServiceName()).orElseThrow(() -> new RuntimeException("No LDAP URL is available."));
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java]: You seem to use route hostname to access LDAP. Does it work? I thought that OpenShift route is able to handle just HTTP and HTTPS communication.

Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java]: Yes, this is working fine.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java [hunk 4]
```diff
@@ -255,34 +256,20 @@ public void setResources(Map<String, String> requests, Map<String, String> limit
                  .done();
     }
 
+    protected Optional<String> getLdapRoute(String serviceName) {
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java]: Where does this route point to?

Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java]: The route will be something like ldap://ldap-${APPLICATION_NAME}.project.openshiftdomain:30389

This is according to the route name I configured in the ldap.yaml template.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    public ExternalDeploymentID getKey() {
+        return ExternalDeploymentID.LDAP;
+    }
+
+    @Override
+    protected LdapDeployment deployToProject(Project project) {
+        logger.info("Creating internal LDAP instance.");
+        project.processTemplateAndCreateResources(getLdapTemplate(), getLdapProperties(project));
+
+        logger.info("Waiting for LDAP deployment to become ready.");
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java]: Seems to be a leftover, I don't see any waiting code in here.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    private static CloudProperties INSTANCE;
+
+    private final Properties prop;
+
+    private CloudProperties() {
+        String fileLocation = System.getProperty(CLOUD_PROPERTIES_LOCATION);
+        try (InputStream is = new FileInputStream(new File(fileLocation))) {
+            prop = new Properties();
+            prop.load(is);
+        } catch (Exception e) {
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Can be replaced by IOException

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java (hunk 4 above)
Comment (sutaakar) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java]: But AFAIK the route in template is available at http://ldap-${APPLICATION_NAME}.project.openshiftdomain:80
As you specify there just hostname, protocol and port are added implicitly.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java (hunk 1 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Fixed

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java (hunk 3 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java]: This has been reverted out.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java (hunk 4 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/OpenShiftDeployment.java]: This has been reverted out.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java (hunk 5 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java]: My bad. Removed.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java (hunk 6 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Replaced

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    public String getLdapDockerImage() {
+        return getProperty(LDAP_DOCKER_IMAGE_PROPERTY);
+    }
+
+    public String getProperty(String key) {
+        return prop.getProperty(key);
+    }
+
+    public static synchronized final CloudProperties getInstance() {
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Can you please switch synchronized and final ? :)

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            return new URL("file://" + getClass().getResource(LDAP_TEMPLATE).getFile());
+        } catch (MalformedURLException e) {
+            throw new RuntimeException("Wrong LDAP template location", e);
+        }
+    }
+
+    private Map<String, String> getLdapProperties(Project project) {
+        Map<String, String> properties = new HashMap<>();
+        properties.put("APPLICATION_NAME", project.getName());
+        properties.put("LDAP_DOCKER_IMAGE", CloudProperties.getInstance().getLdapDockerImage());
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java]: This property doesn't seem to be used, can you please adjust template to use this image tag?

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public String getDeploymentConfigName() {
+        return DEPLOYMENT_CONFIG_NAME;
+    }
+
+    @Override
+    public String getHost() {
+        if (host == null) {
+            host = String.format(HOST_TEMPLATE, getServiceName());
```
Comment (sutaakar) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java]: It may be better to retrieve ldap port using OpenShift too, to avoid hardcoding values and their duplicity.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java (hunk 7 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: According to the link you shared, this was the expected order ... ?

Comment (sutaakar) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Well, I got a complain from my Sonal lint about modifiers order stating above.
But anyway this is a minor thing, you can keep it like this.

Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Tbh, I felt uncomfortable to set it using this order. I prefer to follow the sonar and checkstyle rule about this, so I will change it, no worries.

Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Done

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java (hunk 8 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/external/impl/AbstractLdapExternalDeployment.java]: My bad. I totally forgot to commit the template. Fixed.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java (hunk 9 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/deployment/LdapDeploymentImpl.java]: I didn't want to overcomplicate this logic, but I changed it as you suggested.

**Comment by Sgitario** [comment]
PR updated with latest comments from @sutaakar

**Comment by sutaakar** [comment]
Looks good to me, I will check it on OCP instance once it is up and running.

File: pom.xml [hunk 10]
```diff
@@ -23,7 +23,7 @@
   </modules>
 
   <properties>
-    <rhba.images.version.prefix>77</rhba.images.version.prefix>
+    <rhba.images.version.prefix>78</rhba.images.version.prefix>
```
Comment (sutaakar) [inline pom.xml]: This should be ideally placed as a separate PR, as it is not directly related to the original purpose of PR.
But in this case you can keep it here.

Comment (Sgitario) [reply pom.xml]: I know. I'm running the jobs in templates and operator to ensure this is working fine before merging. (I updated the BXMS repo as well). I will revert it afterwards.

Comment (sutaakar) [reply pom.xml]: ok

File: test-cloud/test-cloud-remote/src/test/resources/deployments/ldap.yaml [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        insecure: true
+      referencePolicy:
+        type: Local
+- kind: ConfigMap
+  apiVersion: v1
+  metadata:
+    name: ldap-user-scripts
+  data: 
+    addUser.sh: |- 
+      ldapadd -h ldap-${APPLICATION_NAME} -p 30389 -x -D cn=Manager,dc=example,dc=com -w admin << EOF
```
Comment (Sgitario) [inline test-cloud/test-cloud-remote/src/test/resources/deployments/ldap.yaml]: I added this script so we ensure that the user we specify in the test params, has the right credentials in LDAP.

File: pom.xml (hunk 10 above)
Comment (Sgitario) [reply pom.xml]: reverted

File: framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/scenario/builder/ClusteredWorkbenchRuntimeSmartRouterTwoKieServersTwoDatabasesScenarioBuilder.java [hunk 12]
```diff
@@ -58,13 +58,23 @@
...
-     * @param ldapSettings configuration of LDAP represented by a class.
-     * @return Builder
-     */
-    ClusteredWorkbenchRuntimeSmartRouterTwoKieServersTwoDatabasesScenarioBuilder withLdapSettings(LdapSettings ldapSettings);
+    *
+    * Return setup builder with an external LDAP.
+    *
+    * @param ldapSettings configuration of LDAP represented by a class.
+    * @return Builder
+    */
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/scenario/builder/ClusteredWorkbenchRuntimeSmartRouterTwoKieServersTwoDatabasesScenarioBuilder.java]: My IDE is already formatting all the JavaDocs automatically. Also, I did check the formatter and it looks ok to me. Can you check your IDE and Eclipse is behaving the same when formatting these comments?

File: framework-cloud/framework-openshift-apb/src/main/java/org/kie/cloud/openshift/deployment/external/ExternalDeploymentFactoryApb.java [hunk 13]
```diff
@@ -22,8 +22,11 @@
 
 public class ExternalDeploymentFactoryApb implements ExternalDeploymentFactory {
 
+    @Override
     public ExternalDeployment<? extends Deployment, ?> get(ExternalDeploymentID id, Map<String, String> extraDeploymentConfig) {
         switch (id) {
+            case LDAP:
+                return new LdapExternalDeploymentApb(extraDeploymentConfig);
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply framework-cloud/framework-openshift-apb/src/main/java/org/kie/cloud/openshift/deployment/external/ExternalDeploymentFactoryApb.java]: I added this class a few days ago. I guess when APB gets removed, this file among others will be removed as well.

File: framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/scenario/builder/ClusteredWorkbenchRuntimeSmartRouterTwoKieServersTwoDatabasesScenarioBuilder.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply framework-cloud/framework-cloud-api/src/main/java/org/kie/cloud/api/scenario/builder/ClusteredWorkbenchRuntimeSmartRouterTwoKieServersTwoDatabasesScenarioBuilder.java]: I've manually fixed the javadocs for all the new methods.

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            throw new RuntimeException("Error loarding cloud properties", e);
+        }
+    }
+
+    public String getLdapDockerImage() {
+        return getProperty(LDAP_DOCKER_IMAGE_PROPERTY);
+    }
+
+    public String getProperty(String key) {
+        return prop.getProperty(key);
```
Comment (jakubschwan) [inline framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: It would be good to handle NPE that can be thrown when property does not exists in Cloud properties file.
It would be good to catch NPE and throw it as a RuntimeException with some additional info for missing property. @sutaakar wdyt?

File: test-cloud/pom.xml [hunk 15]
```diff
@@ -27,6 +27,7 @@
     <openshift.admin.password>admin</openshift.admin.password>
 
     <kie.image.streams/> <!-- Needs to be defined for proper test run. -->
+    <cloud.properties.location/> <!-- Needs to be defined for proper test run. -->
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply test-cloud/pom.xml]: Actually, the idea is to add more properties in this file instead of adding more and more properties that require to modify the jenkins job. 

I will add in the README.md what properties are expected (by looking in the CloudProperties.java file indeed).

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java (hunk 14 above)
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Done

File: framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+ */
+package org.kie.cloud.openshift.resource;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public final class CloudProperties {
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/resource/CloudProperties.java]: Added some information.
